core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/colsAtAnyDepth.kt

```diff
@@ -0,0 +1,184 @@
+package org.jetbrains.kotlinx.dataframe.api
+
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNotBe
+import org.jetbrains.kotlinx.dataframe.columns.ColumnKind.Frame
+import org.jetbrains.kotlinx.dataframe.columns.ColumnKind.Group
+import org.jetbrains.kotlinx.dataframe.columns.ColumnKind.Value
+import org.jetbrains.kotlinx.dataframe.columns.ColumnSet
+import org.jetbrains.kotlinx.dataframe.columns.ColumnWithPath
+import org.jetbrains.kotlinx.dataframe.columns.asColumnSet
+import org.jetbrains.kotlinx.dataframe.impl.columns.TransformableColumnSet
+import org.jetbrains.kotlinx.dataframe.impl.columns.atAnyDepthImpl
+import org.jetbrains.kotlinx.dataframe.impl.columns.transform
+import org.jetbrains.kotlinx.dataframe.impl.columns.tree.flattenRecursively
+import org.jetbrains.kotlinx.dataframe.kind
+import org.jetbrains.kotlinx.dataframe.samples.api.TestBase
+import org.jetbrains.kotlinx.dataframe.samples.api.city
+import org.jetbrains.kotlinx.dataframe.samples.api.firstName
+import org.jetbrains.kotlinx.dataframe.samples.api.lastName
+import org.jetbrains.kotlinx.dataframe.samples.api.name
+import org.junit.Test
+import kotlin.reflect.typeOf
+
+class AtAnyDepth : TestBase() {
+
+    fun List<ColumnWithPath<*>>.print() {
+        forEach {
+            if (it.isValueColumn()) println("${it.name}: ${it.type()}")
+            else it.print()
+        }
+        println()
+    }
+
+    infix fun List<ColumnWithPath<*>>.shouldBe(other: List<ColumnWithPath<*>>) {
+        this.map { it.name to it.path } shouldBe other.map { it.name to it.path }
+    }
+
+    infix fun List<ColumnWithPath<*>>.shouldNotBe(other: List<ColumnWithPath<*>>) {
+        this.map { it.name to it.path } shouldNotBe other.map { it.name to it.path }
+    }
+
+    // old function copied over to avoid breaking changes
+    private fun ColumnSet<*>.dfsInternal(predicate: (ColumnWithPath<*>) -> Boolean): TransformableColumnSet<Any?> =
+        transform {
+            it.filter { it.isColumnGroup() }
+                .flatMap { it.cols().flattenRecursively().filter(predicate) }
+        }
+
+    private val atAnyDepthGoal =
+        dfGroup.getColumnsWithPaths {
+            asSingleColumn().ensureIsColumnGroup().asColumnSet().dfsInternal { true }
+        }.sortedBy { it.name }
+
+    private val atAnyDepthNoGroups =
+        dfGroup.getColumnsWithPaths {
+            asSingleColumn().ensureIsColumnGroup().asColumnSet().dfsInternal { !it.isColumnGroup() }
+        }.sortedBy { it.name }
+
+    private val atAnyDepthString = dfGroup.getColumnsWithPaths {
+        asSingleColumn()
+            .ensureIsColumnGroup()
+            .asColumnSet()
+            .dfsInternal { it.isSubtypeOf(typeOf<String?>()) }
+    }.sortedBy { it.name }
+
+    @Test
+    fun `first, last, and single`() {
+        listOf(
+            dfGroup.select { name.firstName.firstName },
```

Comment:
could we change a subfield from firstName.firstName to firstName.newFirstName at least?